### ISSUE TITLE
Support contracts with library links

### ISSUE THREAD

**Issue by FabijanC** [ethereum/sourcify#468]
This was already attempted in #105.

- Some time ago a question was raised (https://github.com/ethereum/sourcify/pull/105#issuecomment-648293897) about whether it makes sense to classify as `full match` contracts that contain references to libraries that aren't a `full match` themselves. But is it not the case that, for the contracts to be verified, the sources of the libraries also had to be provided and are then stored in the repository, effectively making them verified?
- If bytecode comparison determines that the difference between the deployed and recompiled bytecode is just in the library address placeholders, can such a contract be classified as a match, be it a `full match` or a `partial match`? That is, if initial comparison determines the mentioned difference, the actual library addresses can be taken from the deployed bytecode and used in a new recompilation+verification process. Then if the newly recompiled bytecode matches the deployed one (even the metadata hash), why not deem it a `full match`?
- Or perhaps after an unsuccessful verification procedure due to the recompiled bytecode containing address placeholders, an improved UI could prompt the user for the missing addresses (but this approach is just UI oriented).
- Regarding what should be stored in case of successful verification, this was already referred to in #112.

**Comment by FabijanC** [comment]
@chriseth What do you think?

**Comment by chriseth** [comment]
We discussed that in the meeting, didn't we?

Referencing libraries by their address should not require any source of the library, but we should do the following:

Library addresses can be directly given to the compiler. In this case, they will be part of the metadata and the regular compilation will check them.

In the the other case, when not all library addresses are provided, the recompiled bytecode will contain placeholders. In that case, we should check the (creation) at the contract's address and extract what the placeholders were replaced by. The procedure should verify that the same library has always been replaced by the same address. If this succeeds, we have a full match and store the association between library and extracted address in a file (similar as what we do with constructor arguments).
